src/sync/sys/windows/net.rs

```diff
@@ -0,0 +1,311 @@
+        }
+    }
+
+    pub(crate) fn id(&self) -> i32 {
+        self.named_pipe as i32
+    }
+
+    pub fn read(&self, buf: &mut [u8]) -> Result<usize> {
+        trace!("starting read for thread {:?} on pipe instance {}", std::thread::current().id(), self.named_pipe as i32);
+        let ol = Overlapped::new_with_event(self.read_event);
+
+        let len = std::cmp::min(buf.len(), u32::MAX as usize) as u32;
+        let mut bytes_read= 0;
+        let result = unsafe { ReadFile(self.named_pipe, buf.as_mut_ptr() as *mut _, len, &mut bytes_read,ol.as_mut_ptr()) };
+        if result > 0 && bytes_read > 0 {
+            // Got result no need to wait for pending read to complete
+            return Ok(bytes_read as usize)
+        }
+
+        // wait for pending operation to complete (thread will be suspended until event is signaled)
+        match io::Error::last_os_error() {
+            ref e if e.raw_os_error() == Some(ERROR_IO_PENDING as i32) => {
+                let mut bytes_transfered = 0;
+                let res = unsafe {GetOverlappedResult(self.named_pipe, ol.as_mut_ptr(), &mut bytes_transfered, WAIT_FOR_EVENT) };
+                match res {
+                    0 => {
+                        return Err(Error::Windows(io::Error::last_os_error().raw_os_error().unwrap()))
+                    }
+                    _ => {
+                        return Ok(bytes_transfered as usize)
+                    }
+                }
+            }
+            ref e => {
+                return Err(Error::Others(format!("failed to read from pipe: {:?}", e)))
+            }
+        }
+    }
+
+    pub fn write(&self, buf: &[u8]) -> Result<usize> {
+        trace!("starting write for thread {:?} on pipe instance {}", std::thread::current().id(), self.named_pipe as i32);
+        let ol = Overlapped::new_with_event(self.write_event);
+        let mut bytes_written= 0;
+        let len = std::cmp::min(buf.len(), u32::MAX as usize) as u32;
+        let result = unsafe { WriteFile(self.named_pipe, buf.as_ptr() as *const _,len, &mut bytes_written, ol.as_mut_ptr())};
+        if result > 0 && bytes_written > 0 {
+            // No need to wait for pending write to complete
+            return Ok(bytes_written as usize)
+        }
+
+        // wait for pending operation to complete (thread will be suspended until event is signaled)
+        match io::Error::last_os_error() {
+            ref e if e.raw_os_error() == Some(ERROR_IO_PENDING as i32) => {
+                let mut bytes_transfered = 0;
+                let res = unsafe {GetOverlappedResult(self.named_pipe, ol.as_mut_ptr(), &mut bytes_transfered, WAIT_FOR_EVENT) };
+                match res {
+                    0 => {
+                        return Err(Error::Windows(io::Error::last_os_error().raw_os_error().unwrap()))
+                    }
+                    _ => {
+                        return Ok(bytes_transfered as usize)
+                    }
```

Comment:
Wonder if we should have a helper for this match pattern given how common it is

Reply:
Had the same thought and tried to move it to something like `process_overlapped()` but when I did things didn't work properly. Given it is only repeated twice (for read and write) I decided to keep it as is.

Reply:
Oh I meant the general 0 or _ match pattern. accept, read, write, shutdown and close_handle follow this trend, but might not be worth the hassle. Probably clearer to have it in the call site